libs/clients/middlewares/src/lib/withAutoAuth.ts

```diff
@@ -0,0 +1,244 @@
+    cache:
+      cache && useCache
+        ? {
+            cacheManager: cache.cacheManager,
+            overrideForPost: true,
+            shared: false,
+            overrideCacheControl: async (req, res) => {
+              if (res.ok) {
+                const response = (await res.clone().json()) as TokenResponse
+                return buildCacheControl({ maxAge: response.expires_in })
+              }
+              return buildCacheControl({ noStore: true })
+            },
+          }
+        : undefined,
+  })
+
+  const getAuth = async (
+    input: RequestInfo,
+    init: RequestInit | undefined,
+  ): Promise<string> => {
+    const auth = init?.auth ?? (input as Request).auth
+    const isTokenExchange = options.mode !== 'token' && !!auth
+
+    if (options.mode === 'tokenExchange' && !auth) {
+      const message = `Fetch failure (${name}): Could not perform token exchange. Auth object was missing.`
+      logger.error({
+        url: input,
+        message,
+      })
+      throw new Error(message)
+    }
+
+    if (!isTokenExchange) {
+      const authorization = await tokenCacheManager.get<string>(TOKEN_CACHE_KEY)
+      if (authorization) {
+        return authorization
+      }
+    }
+
+    // In some cases the original access token has everything we need. Then we can skip a token exchange.
+    const hasAllScopes = options.scope.every((scope) =>
+      auth?.scope.includes(scope),
+    )
+    if (
+      auth &&
+      isTokenExchange &&
+      !alwaysTokenExchange &&
+      hasAllScopes &&
+      (!requestActorToken || !auth.actor)
+    ) {
+      return auth.authorization
+    }
+
+    const params = new URLSearchParams({
+      grant_type: 'client_credentials',
+      client_id: options.clientId,
+      client_secret: options.clientSecret,
+      scope: options.scope.join(' '),
+    })
+
+    if (auth && isTokenExchange) {
+      params.set(
+        'grant_type',
+        'urn:ietf:params:oauth:grant-type:token-exchange',
+      )
+      params.set('subject_token', auth.authorization.replace(/^bearer /i, ''))
+      params.set(
+        'subject_token_type',
+        'urn:ietf:params:oauth:token-type:access_token',
+      )
+      params.set(
+        'requested_token_type',
+        requestActorToken
+          ? 'islandis:oauth:token-type:actor-access-token'
+          : 'urn:ietf:params:oauth:token-type:access_token',
+      )
+    }
+
+    const response = await innerFetch(tokenEndpoint, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+      },
+      body: params,
+      auth,
+    })
+
+    const result: TokenResponse = await response.json()
+
+    if (result.token_type !== 'Bearer') {
+      throw new Error(
+        `Fetch (${name}): Token exchange failed, invalid token type (${result.token_type})`,
+      )
+    }
+
+    if (
+      isTokenExchange &&
+      result.issued_token_type !=
+        'urn:ietf:params:oauth:token-type:access_token'
```

Comment:
Shouldn't the `issued_token_type` match the `requested_token_type`? So if `requestActorToken` is `true` then it would be `islandis:oauth:token-type:actor-access-token`?

Also we should maybe have a enums for `TokenType` and `GrantType` as these uris are not so nice copy-paste around. i.e.

```
export enum TokenType { 
  AccessToken = 'urn:ietf:params:oauth:token-type:access_token',
  ActorAccessToken = 'islandis:oauth:token-type:actor-access-token'
}

export enum GrantType {
  TokenExchane = 'urn:ietf:params:oauth:grant-type:token-exchange',
  ClientCredentials = 'client_credentials',
  AuthorizationCode = 'code'
}
```